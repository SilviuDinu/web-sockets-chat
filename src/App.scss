* {
  box-sizing: border-box;
}

.App {
  text-align: center;
}

.App {
  background-color: #282c34;
  min-height: 500px;
  display: flex;
  flex-direction: column;
  color: #e4e6eb;

  .App-header {
    min-height: 80px;
    margin-top: 32px;
  }

  main {
    margin-top: 40px;
    display: flex;
    justify-content: center;
    align-items: center;
    align-content: center;
    .welcome-area {
      max-width: 500px;
      width: 100%;
      min-width: 100%;
      display: flex;
      justify-content: space-evenly;
      .welcome-area-form {
        display: flex;
        flex-direction: row;
        justify-content: space-evenly;

        input {
          width: 250px;
        }

        .button {
          width: 90px;
        }
      }
    }

    .chat-area {
      background-color: #242526;
      width: 600px;
      min-height: 365px;
      display: flex;
      flex-direction: column;
      justify-content: flex-end;

      @media (max-width: 960px) {
        max-width: 90vw;
      }

      .chat-area-box {
        padding-left: 16px;
        padding-right: 16px;
        max-height: 300px;
        overflow-y: auto;
        .message-list {
          display: flex;
          justify-content: space-between;
          flex-direction: column;

          .sender {
            font-size: 0.6875rem;
            color: #b0b3b8;
            line-height: 1.33;
            margin-bottom: 4px;
            margin-top: 4px;
          }

          .message-out {
            align-self: flex-end;
            display: flex;
            flex-direction: column;
            margin-top: 8px;

            .sender {
              align-self: flex-end;
              margin-right: 16px;
            }

            .message {
              color: #e4e6eb;
              background-color: #1877f2;
              align-items: center;
              align-content: center;
              justify-content: center;
              display: flex;
              font-size: 16px;
              border-radius: 50px;
              height: 40px;
              padding-left: 16px;
              padding-right: 16px;
            }
          }

          .message-in {
            align-self: flex-start;
            display: flex;
            flex-direction: column;
            margin-top: 8px;

            .sender {
              align-self: flex-start;
              margin-left: 16px;
            }

            .message {
              color: #e4e6eb;
              background-color: #3e4042;
              align-items: center;
              align-content: center;
              justify-content: center;
              display: flex;
              font-size: 16px;
              border-radius: 50px;
              height: 40px;
              padding-left: 16px;
              padding-right: 16px;
            }
          }
        }
      }

      .chat-area-form-wrapper {
        padding-bottom: 8px;
        margin-top: 16px;
        display: flex;
        flex-direction: column;

        input {
          width: 70%;
        }

        .button {
          width: 22.5%;
        }

        @media (min-width: 600px) and (max-width: 960px) {
          input {
            width: 70%;
          }
        }

        @media (max-width: 600px) {
          input {
            width: 65%;
          }
        }
      }
    }
  }

  .error {
    color: #ff0033;
  }

  .message {
    color: #ecf3ff;
  }

  input {
    height: 40px;
    padding-left: 24px;
    padding-right: 24px;
    border: 0;
    border-radius: 50px;
    background-color: #3a3b3c;
    margin: 0 8px;
    font-size: 16px;
    color: #ecf3ff;
    outline: none;
  }

  .button {
    height: 40px;
    min-width: 60px;
    border: 0;
    border-radius: 50px;
    margin: 0 8px;
    background-color: #2851a3;
    color: #ecf3ff;
    font-size: 16px;
    outline: none;
    cursor: pointer;

    &:hover {
      background-color: #1877f2;
    }
  }
}
