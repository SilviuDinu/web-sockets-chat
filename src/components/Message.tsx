import { useEffect, useRef } from 'react';

export default function Message(props: any) {
  const chatBoxRef = useRef<any>();

  const updateScrollPos = (): void => {
    chatBoxRef.current.scrollIntoView(true);
    const element = document.querySelector<any>('.chat-area .chat-area-box');
    if(element) {
      element.scrollTop += 10;
    }
  };

  useEffect(() => {
    if (chatBoxRef && chatBoxRef.current) {
      updateScrollPos();
    }
  });

  return (
    <div className={props.type}>
      <div className="sender">{props.sender}</div>
      <div className="message">
        <span ref={chatBoxRef}>{props.message}</span>
      </div>
    </div>
  );
}
