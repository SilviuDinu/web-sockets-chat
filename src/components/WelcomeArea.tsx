import Notification from './Notification';
import Form from './Form';

export default function WecomeArea(props: any) {
  const form = {
    input: {
      placeholder: 'Please enter your name...',
      ariaLabel: 'Enter a name to be used in the chat room',
      class: 'welcome-area-form-input',
    },
    buttonText: 'Send',
    class: "welcome-area-form"
  };

  const onInputChange = (e: Event) => {
    props.onNameChange(e);
  }

  const onFormSubmit = (e: Event) => {
      props.onNameSubmit(e);
  }

  return props.isVisible ? (
    <>
      <div className="welcome-area">
        <Form
            form={form}
            onInputChange={onInputChange}
            onFormSubmit={onFormSubmit}>
        </Form>
      </div>
      {props.hasAnyConnectionError ? <Notification type="error" elem="p" message={props.notification.message} /> : null}
    </>
  ) : null;
}
