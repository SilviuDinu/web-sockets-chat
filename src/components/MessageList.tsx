import Message from './Message';

export default function MessageList(props: any) {
    return props.messages.length > 0 ? (
      <div className="message-list">
        {props.messages.map((item: any, index: number) => {
            return (
              <Message
                key={item.id || index}
                type={item.type}
                sender={item.name || null}
                message={item.message} />
              );
        })}
      </div>
    ): null;
  }