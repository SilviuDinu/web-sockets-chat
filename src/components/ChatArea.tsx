import MessageList from './MessageList';
import Notification from './Notification';
import Form from './Form';

export default function ChatArea(props: any) {

  const form = {
    input: {
      placeholder: 'Write a message ...',
      ariaLabel: 'Enter message to send',
      class: 'chat-area-form-input',
    },
    buttonText: 'Send',
    class: 'chat-area-form',
  };

  const onInputChange = (e: Event) => {
    props.onMessageChange(e);
  };

  const onFormSubmit = (e: Event) => {
    props.onMessageSend(e);
  };

  return props.isVisible ? (
    <div className="chat-area">
      <div className="chat-area-box">
        <MessageList messages={props.chat} />
        {props.typing.enabled ? <Notification
          type="temporary"
          elem="span"
          message={props.typing.message} /> : null }
      </div>
      <div className="chat-area-form-wrapper">
        <Form
            form={form}
            inputValue={props.messageOut}
            onInputChange={onInputChange}
            onFormSubmit={onFormSubmit}>
        </Form>
      </div>
    </div>
  ) : null;
}
