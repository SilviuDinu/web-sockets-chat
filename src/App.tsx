import { useEffect, useState } from 'react';
import { io } from 'socket.io-client';

import './App.scss';
import Header from './components/Header';
import Body from './components/Body';
import WelcomeArea from './components/WelcomeArea';
import ChatArea from './components/ChatArea';

import { MESSAGES } from './data/enums/messages.enum';
import { Notification } from './data/models/notification.model';

const socket = io('/', { transports: ['polling'] });

function App() {
  const [title, setTitle] = useState<MESSAGES | string>(MESSAGES.TITLE);
  const [connected, setConnected] = useState<boolean>(false);
  const [name, setName] = useState<string>('');
  const [messageOut, setMessageOut] = useState<string>('');
  const [chat, setChat] = useState<any[]>([]);
  const [notification, setNotification] = useState<Notification>({ message: '', type: 'message' });
  const [hasAnyConnectionError, setHasAnyConnectionError] = useState<boolean>(false);

  useEffect(() => {
    socket.on('message', ({ message, sender }: any) => {
      setChat(chat => [
        ...chat,
        {
          id: chat.length + 1,
          name: sender,
          message,
          type: 'message-in',
        },
      ]);
    });
  }, []);

  const joinRoom = (e: Event) => {
    if (e) {
      e.preventDefault();
    }
    if (!connected) {
      socket.emit('join room', { name });
      socket.on('connection success', (success: string) => {
        setConnected(true);
        setTitle(MESSAGES.WELCOME + name);
        setNotification({
          message: success,
          type: 'message',
        });
        setHasAnyConnectionError(false);
      });
      socket.on('connection error', (err: string) => {
        setConnected(false);
        setNotification({
          message: err,
          type: 'error',
        });
        setHasAnyConnectionError(true);
      });
    }
  };

  const sendMessage = (e: Event) => {
    if (e) {
      e.preventDefault();
    }
    if (connected) {
      handleMessageOut();
    }
  };

  const handleMessageOut = (): void => {
    const out = {
      id: chat.length + 1,
      name,
      message: messageOut,
      type: 'message-out',
    };
    socket.emit('send message', out);
    setChat([
      ...chat,
      out
    ]);
    setMessageOut('');
  };

  const onNameChange = (e: Event): void => {
    setName((e.target as HTMLTextAreaElement).value || '');
  };

  const onMessageChange = (e: Event): void => {
    setMessageOut((e.target as HTMLTextAreaElement).value || '');
  };

  return (
    <div className="App">
      <Header title={title}></Header>
      <Body>
        <WelcomeArea
          isVisible={!connected}
          name={name}
          onNameSubmit={joinRoom}
          onNameChange={onNameChange}
          hasAnyConnectionError={hasAnyConnectionError}
          notification={notification}>
        </WelcomeArea>
        <ChatArea
          isVisible={connected}
          chat={chat}
          messageOut={messageOut}
          onMessageChange={onMessageChange}
          onMessageSend={sendMessage}>
        </ChatArea>
      </Body>
    </div>
  );
}

export default App;
