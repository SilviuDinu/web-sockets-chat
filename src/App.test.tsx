import React, { useState } from 'react';
import { render, screen } from '@testing-library/react';
import { io } from 'socket.io-client';
import App from './App';
import { MESSAGES } from './data/enums/messages.enum';
import WelcomeArea from './components/WelcomeArea';
import MessageList from './components/MessageList';
import Message from './components/Message';
import ChatArea from './components/ChatArea';
import Header from './components/Header';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

const MessageMock = (props: any) => {
  return (
    <div className={props.type}>
      <div className="sender">{props.sender}</div>
      <div className="message">
        <span>{props.message}</span>
      </div>
    </div>
  );
};

test(`renders Silviu's Chat Application`, () => {
  render(<App />);
  const linkElement = screen.getByText(/Silviu's Chat Application/i);
  expect(linkElement).toBeInTheDocument();
});

test(`renders Header`, () => {
  const title = MESSAGES.WELCOME + 'William';
  render(<Header title={title} />);
  const welcomeMessage = screen.getByText('Welcome, William');
  expect(welcomeMessage).toBeInTheDocument();
});

test(`renders Welcome Area`, () => {
  const component = renderer.create(
    <WelcomeArea
      isVisible={true}
      name={'William'}
      onNameSubmit={() => {}}
      onNameChange={() => {}}
      hasAnyConnectionError={true}
      notification={'eroare'}></WelcomeArea>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test(`renders Chat Area`, () => {
  const component = renderer.create(
    <ChatArea
      isVisible={true}
      chat={[
        {
          id: 0,
          message: 'William has joined the chat',
          type: 'notification',
        },
        {
          id: 1,
          message: 'John has joined the chat',
          type: 'notification',
        },
      ]}
      typing={false}
      messageOut={'Salut'}
      onMessageChange={() => {}}
      onMessageSend={() => {}}></ChatArea>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test(`test MessageList`, () => {
  const component = renderer.create(
    <MessageList
      messages={[
        {
          id: 0,
          message: 'Salut',
          type: 'message-in',
        },
      ]}></MessageList>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test(`test Message`, () => {
  window.HTMLElement.prototype.scrollIntoView = function () {
    document.querySelector<any>('body').scrollTop += 10;
  };

  render(
    <MessageList
      messages={[
        {
          id: 0,
          message: 'Salut',
          type: 'message-in',
        },
      ]}></MessageList>
  );
  const msg = screen.getByText(/Salut/i);
  expect(msg).toBeInTheDocument();
});

jest.mock('socket.io-client', () => {
  const emit = jest.fn();
  const on = jest.fn();
  const socket = { emit, on };
  return jest.fn(() => socket);
});

