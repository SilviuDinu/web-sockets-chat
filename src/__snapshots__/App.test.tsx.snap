// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Chat Area 1`] = `
<div
  className="chat-area"
>
  <div
    className="chat-area-box"
  >
    <div
      className="message-list"
    >
      <div
        className="notification"
      >
        <div
          className="sender"
        />
        <div
          className="message"
        >
          <span>
            William has joined the chat
          </span>
        </div>
      </div>
      <div
        className="notification"
      >
        <div
          className="sender"
        />
        <div
          className="message"
        >
          <span>
            John has joined the chat
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="chat-area-form-wrapper"
  >
    <form
      className="chat-area-form"
      method="POST"
      onSubmit={[Function]}
    >
      <input
        aria-label="Enter message to send"
        autoComplete="off"
        className="chat-area-form-input"
        name="name"
        onChange={[Function]}
        placeholder="Write a message ..."
        type="text"
        value="Salut"
      />
      <button
        className="button"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</div>
`;

exports[`renders Welcome Area 1`] = `
Array [
  <div
    className="welcome-area"
  >
    <form
      className="welcome-area-form"
      method="POST"
      onSubmit={[Function]}
    >
      <input
        aria-label="Enter a name to be used in the chat room"
        autoComplete="off"
        className="welcome-area-form-input"
        name="name"
        onChange={[Function]}
        placeholder="Please enter your name..."
        type="text"
        value="William"
      />
      <button
        className="button"
        type="submit"
      >
        Send
      </button>
    </form>
  </div>,
  <div
    className="error"
  >
    <p />
  </div>,
]
`;

exports[`test MessageList 1`] = `
<div
  className="message-list"
>
  <div
    className="message-in"
  >
    <div
      className="sender"
    />
    <div
      className="message"
    >
      <span>
        Salut
      </span>
    </div>
  </div>
</div>
`;
